### PR TITLE
Add unit test case for version adaptibility

### PR DESCRIPTION
**Pull request by sinhaashish**

### CODE REVIEW

**Comment by Praveenrajmani** [comment]
Can we add some steps documented in https://github.com/minio/direct-csi/blob/master/docs/development-and-testing.md @balamurugana

File: pkg/client/drive_adapter_test.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	)
+	directcsidrivesKind := schema.GroupVersionKind{Group: "direct.csi.min.io", Version: backendVersion, Kind: "DirectCSIDrive"}
+	return &directCSIDriveAdapter{resourceInterface: dynamicResourceClient, gvk: &directcsidrivesKind}, nil
+}
+
+func TestGet(t *testing.T) {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+
+	l := newUnstructured("group/version", "TheKind", "ns-foo", "name-foo")
```
Comment (Praveenrajmani) [inline pkg/client/drive_adapter_test.go:66]: you need to provide DirectCSIDrive's g.v.k here @sinhaashish

**Comment by Praveenrajmani** [comment]
For watch, we can implement a wrapper around the existing watch interface, Like the following, 

```
type watchInterfaceWrapper struct {
	watchInterface watch.Interface
}

func (w watchInterfaceWrapper) Stop() {
	w.watchInterface.Stop()
}

func (w watchInterfaceWrapper) ResultChan() <-chan watch.Event {
	wrapperCh := make(chan watch.Event)
	go func() {
		defer close(wrapperCh)
		resultCh := w.watchInterface.ResultChan()
		for {
			result, ok := <-resultCh
			if !ok {
				break
			}
			convertedObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&result.Object)
			if err != nil {
				break
			}
			intermediateResult := &unstructured.Unstructured{Object: convertedObj}
			finalResult := &unstructured.Unstructured{}
			if err := converter.Migrate(intermediateResult, finalResult, schema.GroupVersion{
				Version: directcsi.Version,
				Group:   directcsi.Group,
			}); err != nil {
				klog.Info("error while migrating: %v", err)
				break
			}
			var driveObject runtime.Object
			err = runtime.DefaultUnstructuredConverter.FromUnstructured(finalResult.Object, &driveObject)
			if err != nil {
				klog.Info("error while converting from unstructured: %v", err)
				break
			}
			wrapperCh <- watch.Event{
				Type:   result.Type,
				Object: driveObject,
			}
		}
	}()
	return wrapperCh
}

// Watch returns a watch.Interface that watches the requested directCSIDrives.
func (c *directCSIDriveAdapter) Watch(ctx context.Context, opts metav1.ListOptions) (watch.Interface, error) {
	opts.Watch = true
	watcher, err := c.resourceInterface.Watch(ctx, opts)
	if err != nil {
		return nil, err
	}
	return watchInterfaceWrapper{watchInterface: watcher}, nil
}
```

File: pkg/client/drive_adapter_test.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+	)
+	directcsidrivesKind := schema.GroupVersionKind{Group: "direct.csi.min.io", Version: backendVersion, Kind: "DirectCSIDrive"}
+	return &directCSIDriveAdapter{resourceInterface: dynamicResourceClient, gvk: &directcsidrivesKind}, nil
+}
+
+func GetFakeDirectCSIDriveListAdapter() (clientset.DirectCSIDriveInterface, error) {
+
+	fakeResourceInterface := fakedynamic.NewSimpleDynamicClientWithCustomListKinds(runtime.NewScheme(),
+		map[schema.GroupVersionResource]string{
+			{Group: directcsi.Group, Version: "v1beta1", Resource: "directcsidrives"}: "TheKindList",
```
Comment (Praveenrajmani) [inline pkg/client/drive_adapter_test.go:67]: `TheKindList` is invalid - It has to be `DirectCSIDriveList`

File: pkg/client/interface.go [hunk 3]
```diff
@@ -170,10 +171,60 @@ func (d *directCSIInterface) List(ctx context.Context, opts metav1.ListOptions)
...
+			err = runtime.DefaultUnstructuredConverter.FromUnstructured(finalResult.Object, &driveObject)
+			if err != nil {
+				klog.Infof("error while converting from unstructured: %v", err)
+				break
+			}
+			wrapperCh <- watch.Event{
+				Type:   result.Type,
+				Object: driveObject,
+			}
+		}
```
Comment (Praveenrajmani) [inline pkg/client/interface.go:215]:
```diff
diff --git a/pkg/client/interface.go b/pkg/client/interface.go
index 1326f19..97ca874 100644
--- a/pkg/client/interface.go
+++ b/pkg/client/interface.go
@@ -202,15 +202,9 @@ func (w watchInterfaceWrapper) ResultChan() <-chan watch.Event {
                                klog.Infof("error while migrating: %v", err)
                                break
                        }
-                       var driveObject runtime.Object
-                       err = runtime.DefaultUnstructuredConverter.FromUnstructured(finalResult.Object, &driveObject)
-                       if err != nil {
-                               klog.Infof("error while converting from unstructured: %v", err)
-                               break
-                       }
                        wrapperCh <- watch.Event{
                                Type:   result.Type,
-                               Object: driveObject,
+                               Object: finalResult,
                        }
                }
        }()
```

File: pkg/client/drive_adapter_test.go [hunk 4]
```diff
@@ -0,0 +1,693 @@
...
+		if err != nil {
+			t.Errorf("case %v: Error in Create, Expected  err to nil, got %v", i+1, err)
+		}
+
+		_, err = fakeDirectCSIDriveAdapter.Get(ctx, testCase.newDriveName, metav1.GetOptions{})
+		if err != nil {
+			t.Errorf("case %v: Error while fetching newly created drive, Expected  err to nil, got %v", i+1, err)
+		}
+	}
+}
```
Comment (Praveenrajmani) [inline pkg/client/drive_adapter_test.go:693]:
```
func TestWatcher(t *testing.T) {
	ctx, cancelFunc := context.WithCancel(context.Background())
	defer cancelFunc()

	checkEvent := func(event watch.Event, expectedEventType watch.EventType, expectedObject *directcsi.DirectCSIDrive) error {
		if event.Type != expectedEventType {
			return fmt.Errorf("Expected event type: %v, Got: %v", expectedEventType, event.Type)
		}
		var directCSIDrive directcsi.DirectCSIDrive
		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(event.Object.(*unstructured.Unstructured).Object, &directCSIDrive); err != nil {
			fmt.Errorf("Error while converting event object %v", err)
		}
		if !reflect.DeepEqual(&directCSIDrive, expectedObject) {
			fmt.Errorf("invalid object received in the %s event. Expected: %v, Got: %v", expectedEventType, expectedObject, directCSIDrive)
		}
		return nil
	}

	fakeDriveClient, err := getFakeDirectCSIAdapter("v1beta2", "directcsidrives", newUnstructured("v1beta2", "DirectCSIDrive", "drive-1"))
	if err != nil {
		t.Errorf("Error in creating fake drive interface: %v", err)
	}
	fakeWatchInterface, err := fakeDriveClient.Watch(ctx, metav1.ListOptions{})
	if err != nil {
		t.Errorf("Error in creating watch interface: %v", err)
	}
	resultCh := watchInterfaceWrapper{fakeWatchInterface}.ResultChan()
	fakeDirectCSIDriveAdapter := &directCSIDriveInterface{*fakeDriveClient}

	// Test create event
	testCreateObject := &directcsi.DirectCSIDrive{
		TypeMeta: utils.DirectCSIDriveTypeMeta(),
		ObjectMeta: metav1.ObjectMeta{
			Name: "drive-2",
		},
	}
	_, err = fakeDirectCSIDriveAdapter.Create(ctx, testCreateObject, metav1.CreateOptions{})
	if err != nil {
		t.Errorf("Error while fetching newly created drive, Expected  err to nil, got %v", err)
	}
	event, ok := <-resultCh
	if !ok {
		t.Fatalf("no event received for create")
	}
	if err := checkEvent(event, watch.Added, testCreateObject); err != nil {
		t.Fatalf(err.Error())
	}

	// Test Update Event
	testUpdateObject := &directcsi.DirectCSIDrive{
		TypeMeta: utils.DirectCSIDriveTypeMeta(),
		ObjectMeta: metav1.ObjectMeta{
			Name: "drive-1",
		},
		Status: directcsi.DirectCSIDriveStatus{
			DriveStatus: directcsi.DriveStatusReleased,
		},
	}
	_, err = fakeDirectCSIDriveAdapter.Update(ctx, testUpdateObject, metav1.UpdateOptions{})
	if err != nil {
		t.Fatalf("Error while updating fake drive object: %v", err)
	}
	event, ok = <-resultCh
	if !ok {
		t.Fatalf("no event received for update")
	}
	if err := checkEvent(event, watch.Modified, testUpdateObject); err != nil {
		t.Fatalf(err.Error())
	}

	// Test Delete Event
	err = fakeDirectCSIDriveAdapter.Delete(ctx, "drive-1", metav1.DeleteOptions{})
	if err != nil {
		t.Errorf("error while deleting fake object: %v", err)
	}
	event, ok = <-resultCh
	if !ok {
		t.Fatalf("no event received for delete")
	}
	if err := checkEvent(event, watch.Deleted, testUpdateObject); err != nil {
		t.Fatalf(err.Error())
	}
}
```

**Review by Praveenrajmani — COMMENTED** [review]
Some simplifications on the watch interface

File: pkg/client/drive_adapter_test.go [hunk 5]
```diff
@@ -0,0 +1,693 @@
...
+			backendVersion: "v1alpha1",
+			kind:           "DirectCSIDrive",
+			name:           "drive-1",
+			expectedDrive: &directcsi.DirectCSIDrive{
+				TypeMeta: utils.DirectCSIDriveTypeMeta(),
+				ObjectMeta: metav1.ObjectMeta{Name: "drive-1", Labels: map[string]string{
+					"direct.csi.min.io/access-tier": "Unknown",
+					"direct.csi.min.io/created-by":  "directcsi-driver",
+					"direct.csi.min.io/node":        "",
+					"direct.csi.min.io/path":        "",
```
Comment (Praveenrajmani) [inline pkg/client/drive_adapter_test.go:104]: We can test with non-empty label values, The labels will be populated based on this logic - https://github.com/minio/direct-csi/blob/master/pkg/converter/drive_upgrade.go#L149

These labels will be populated when the fields like `Status.NodeName`, `Status.Path` is set on the drives

File: pkg/client/drive_adapter_test.go [hunk 6]
```diff
@@ -0,0 +1,693 @@
...
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "newdrive-3",
+				},
+			},
+		},
+	}
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	for i, testCase := range testCases {
+		inter, err := getFakeDirectCSIAdapter(testCase.backendVersion, "directcsidrives", newUnstructured(testCase.backendVersion, testCase.kind, testCase.name))
```
Comment (Praveenrajmani) [inline pkg/client/drive_adapter_test.go:346]: the variable names can be improved here - `inter` can be `fakeDirecCSIClient`

File: pkg/client/drive_adapter_test.go [hunk 7]
```diff
@@ -0,0 +1,693 @@
...
+
+		err = fakeDirectCSIDriveAdapter.Delete(ctx, testCase.name, metav1.DeleteOptions{})
+		if err != nil {
+			t.Errorf("case %v: Error in Delete, Expected  err to nil, got %v", i+1, err)
+		}
+		_, err = fakeDirectCSIDriveAdapter.Get(ctx, testCase.name, metav1.GetOptions{})
+
+		if err != nil && err.Error() != fmt.Sprintf("%s %q %s", "directcsidrives.direct.csi.min.io", testCase.name, "not found") {
+			t.Errorf("case %v: Error in Get after delete, Expected  err to nil, got %v", i+1, err)
+		}
```
Comment (Praveenrajmani) [inline pkg/client/drive_adapter_test.go:403]: We can use https://pkg.go.dev/k8s.io/apimachinery/pkg/api/errors#IsNotFound to check for not found errors

File: pkg/client/drive_adapter_test.go [hunk 8]
```diff
@@ -0,0 +1,693 @@
...
+		if err != nil {
+			t.Errorf("case %v: Error in creating fake client, Expected err to nil, got %v", i+1, err)
+		}
+		fakeDirectCSIDriveAdapter := &directCSIDriveInterface{*inter}
+		err = fakeDirectCSIDriveAdapter.DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{})
+		if err != nil {
+			t.Errorf("case %v: Error in Delete, Expected  err to nil, got %v", i+1, err)
+		}
+		_, err = fakeDirectCSIDriveAdapter.Get(ctx, testCase.name, metav1.GetOptions{})
+		if err != nil && err.Error() == fmt.Sprintf("%s %q %s", "directcsidrives.direct.csi.min.io", testCase.name, "not found") {
```
Comment (Praveenrajmani) [inline pkg/client/drive_adapter_test.go:442]: https://pkg.go.dev/k8s.io/apimachinery/pkg/api/errors#IsNotFound

**Review by Praveenrajmani — COMMENTED** [review]
The same comments applies to volume adapter tests as well

We can also bring all these tests to a single file and name it - `interface_test.go`

File: pkg/client/drive_adapter_test.go [hunk 9]
```diff
@@ -0,0 +1,693 @@
...
+	directcsi "github.com/minio/direct-csi/pkg/apis/direct.csi.min.io/v1beta3"
+	"github.com/minio/direct-csi/pkg/utils"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	fakedynamic "k8s.io/client-go/dynamic/fake"
+)
+
+func newUnstructured(backendVersion, kind, name string) *unstructured.Unstructured {
```
[the remark this answers is not in the document]
Comment (sinhaashish) [reply pkg/client/drive_adapter_test.go]: with `newUnstructured`  i am passing name as well.

File: pkg/client/drive_adapter_test.go [hunk 10]
```diff
@@ -0,0 +1,693 @@
...
+func TestListDrive(t *testing.T) {
+	testCases := []struct {
+		backendVersion    string
+		kind              string
+		name              []string
+		expectedDriveList *directcsi.DirectCSIDriveList
+	}{
+		{
+			backendVersion: "v1alpha1",
+			kind:           "DirectCSIDrive",
```
[the remark this answers is not in the document]
Comment (sinhaashish) [reply pkg/client/drive_adapter_test.go]: By this format i am passing `DirectCSIDrive` and `DirectCSIVolume` in kind

**Review by wlan0 — APPROVED** [review]

**Pull request merged**